### PR TITLE
Add shopping cart

### PR DESCRIPTION
**Pull request by Dimagaa**

### CODE REVIEW

File: src/main/java/com/app/onlinebookstore/model/CartItem.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
+    @JoinColumn(name = "shopping_cart_id", nullable = false)
+    private ShoppingCart shoppingCart;
+    @ToString.Exclude
+    @EqualsAndHashCode.Exclude
+    @ManyToOne
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/model/CartItem.java]: make it lazy

File: src/main/java/com/app/onlinebookstore/security/JwtUtil.java [hunk 2]
```diff
@@ -33,19 +37,23 @@ public String generateToken(String userName) {
 
     public boolean isValidToken(String token) {
         try {
-            Date expirationDate = getClaimFromToken(token, Claims::getExpiration);
-            return new Date().before(expirationDate);
+            parseToken(token);
+            return true;
         } catch (JwtException | IllegalArgumentException e) {
-            throw new JwtException("Expired or invalid JWT token");
+            logger.error("Can't parse jwt: ", e);
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/security/JwtUtil.java]:
```suggestion
            logger.warn("Expired or invalid JWT token");
```

File: src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    @Override
+    public ShoppingCartDto getShoppingCart() {
+        return findShoppingCartByAuthenticatedUser()
+                .map(shoppingCartMapper::toDto)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Shopping cart was not found for the current user"
+                ));
+    }
+
+    @Override
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:40]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+                ));
+        CartItem cartItem = shoppingCart.getCartItems().stream()
+                .filter(item -> request.bookId().equals(item.getBook().getId()))
+                .peek(item -> item.setQuantity(item.getQuantity() + request.quantity()))
+                .findFirst()
+                .orElse(createNewCartItem(shoppingCart, request));
+        return cartItemMapper.toDto(cartItemRepository.save(cartItem));
+    }
+
+    @Override
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/ShoppingCartServiceImpl.java:53]:
```suggestion
    @Override
    @Transactional
```

File: src/main/java/com/app/onlinebookstore/service/impl/UserServiceImpl.java [hunk 5]
```diff
@@ -32,6 +36,17 @@ public UserDto register(UserRegisterRequestDto userRegisterRequestDto)
         user.setPassword(passwordEncoder.encode(user.getPassword()));
         Role role = roleRepository.getRoleByName(Role.RoleName.ROLE_USER);
         user.setRoles(Set.of(role));
-        return userMapper.toDto(userRepository.save(user));
+        User savedUser = userRepository.save(user);
+        ShoppingCart shoppingCart = new ShoppingCart();
```
Comment (vovasalyha) [inline src/main/java/com/app/onlinebookstore/service/impl/UserServiceImpl.java]: I suggest to extract a logic related to shopping cart to a private method

**Review by vovasalyha — CHANGES_REQUESTED** [review]
GJ. Let's improve it a bit

File: src/main/java/com/app/onlinebookstore/controller/ShoppingCartController.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @Operation(
+            summary = "Update a cart item by ID",
+            description = "Update an existing cart item in the user's shopping cart"
+    )
+    @PreAuthorize("hasRole('USER')")
+    @PutMapping("/cart-items/{cartItemId}")
+    public CartItemDto updateCartItem(@PathVariable Long cartItemId,
+                               @RequestBody CartItemUpdateRequestDto request) {
```
Comment (andrii-hoienko) [inline src/main/java/com/app/onlinebookstore/controller/ShoppingCartController.java]:
```suggestion
                               @RequestBody @Valid CartItemUpdateRequestDto request) {
```

File: src/main/java/com/app/onlinebookstore/model/CartItem.java (hunk 1 above)
Comment (andrii-hoienko) [reply src/main/java/com/app/onlinebookstore/model/CartItem.java]: not resolved

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
GJ, I left minor comments

**Review by vovasalyha — APPROVED** [review]

**Pull request merged**